TestSvcUpdateAnnotation is flaky

**Describe the bug**
```text
--- FAIL: TestSvcUpdateAnnotation (20.00s)
Error:     nplagent_test.go:166: successfully created Service: &Service{ObjectMeta:{pod-svc  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[nodeportlocal.antrea.io/enabled:true] [] []  []},Spec:ServiceSpec{Ports:[]ServicePort{ServicePort{Name:,Protocol:TCP,Port:80,TargetPort:{0 80 },NodePort:0,AppProtocol:nil,},},Selector:map[string]string{foo: pod-svc,},ClusterIP:,Type:ClusterIP,ExternalIPs:[],SessionAffinity:,LoadBalancerIP:,LoadBalancerSourceRanges:[],ExternalName:,ExternalTrafficPolicy:,HealthCheckNodePort:0,PublishNotReadyAddresses:false,SessionAffinityConfig:nil,IPFamily:nil,TopologyKeys:[],},Status:ServiceStatus{LoadBalancer:LoadBalancerStatus{Ingress:[]LoadBalancerIngress{},},},}
Error:     nplagent_test.go:173: successfully created Pod: &Pod{ObjectMeta:{pod-svc  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[foo:pod-svc] map[] [] []  []},Spec:PodSpec{Volumes:[]Volume{},Containers:[]Container{Container{Name:,Image:,Command:[],Args:[],WorkingDir:,Ports:[]ContainerPort{ContainerPort{Name:,HostPort:0,ContainerPort:80,Protocol:,HostIP:,},},Env:[]EnvVar{},Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[]VolumeMount{},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,},},RestartPolicy:,TerminationGracePeriodSeconds:nil,ActiveDeadlineSeconds:nil,DNSPolicy:,NodeSelector:map[string]string{},ServiceAccountName:,DeprecatedServiceAccount:,NodeName:test-node,HostNetwork:false,HostPID:false,HostIPC:false,SecurityContext:nil,ImagePullSecrets:[]LocalObjectReference{},Hostname:,Subdomain:,Affinity:nil,SchedulerName:,InitContainers:[]Container{},AutomountServiceAccountToken:nil,Tolerations:[]Toleration{},HostAliases:[]HostAlias{},PriorityClassName:,Priority:nil,DNSConfig:nil,ShareProcessNamespace:nil,ReadinessGates:[]PodReadinessGate{},RuntimeClassName:nil,EnableServiceLinks:nil,PreemptionPolicy:nil,Overhead:ResourceList{},TopologySpreadConstraints:[]TopologySpreadConstraint{},EphemeralContainers:[]EphemeralContainer{},},Status:PodStatus{Phase:,Conditions:[]PodCondition{},Message:,Reason:,HostIP:10.10.10.10,PodIP:192.168.32.10,StartTime:<nil>,ContainerStatuses:[]ContainerStatus{},QOSClass:,InitContainerStatuses:[]ContainerStatus{},NominatedNodeName:,PodIPs:[]PodIP{},EphemeralContainerStatuses:[]ContainerStatus{},},}
    nplagent_test.go:176: 
        	Error Trace:	nplagent_test.go:176
        	Error:      	Expected nil, but got: &errors.errorString{s:"timed out waiting for the condition"}
        	Test:       	TestSvcUpdateAnnotation
        	Messages:   	Poll for annotation check failed
```

Not sure if this is an issue with the test or the implementation.

**To Reproduce**
This was observed in the Github CI. But I am hoping it can be reproduced locally by running the test enough times.

**Versions:**
 - Antrea version (Docker image tag): top-of-tree

--- a/pkg/agent/nodeportlocal/nplagent_test.go
+++ b/pkg/agent/nodeportlocal/nplagent_test.go
@@ -39,7 +39,7 @@ import (
 	"github.com/vmware-tanzu/antrea/pkg/signals"
 )
 
-func NewPortTable(c *gomock.Controller) *portcache.PortTable {
+func newPortTable(c *gomock.Controller) *portcache.PortTable {
 	ptable := portcache.PortTable{StartPort: 40000, EndPort: 45000}
 	ptable.Table = make(map[int]portcache.NodePortData)
 
@@ -117,7 +117,7 @@ func TestMain(t *testing.T) {
 	os.Setenv("NODE_NAME", defaultNodeName)
 	kubeClient = k8sfake.NewSimpleClientset()
 	mockCtrl := gomock.NewController(t)
-	portTable = NewPortTable(mockCtrl)
+	portTable = newPortTable(mockCtrl)
 
 	// informerFactory is initialised and started from cmd/antrea-agent/agent.go
 	informerFactory := informers.NewSharedInformerFactory(kubeClient, resyncPeriod)
@@ -127,6 +127,11 @@ func TestMain(t *testing.T) {
 
 	go c.Run(stopCh)
 	informerFactory.Start(stopCh)
+
+	// Must wait for cache sync, otherwise resource creation events will be missing if the resources are created
+	// in-between list and watch call of an informer. This is because fake clientset doesn't support watching with
+	// resourceVersion. A watcher of fake clientset only gets events that happen after the watcher is created.
+	informerFactory.WaitForCacheSync(stopCh)
 }
 
 func pollForPodAnnotation(r *require.Assertions, podName string, found bool) (string, error) {

--- a/pkg/agent/nodeportlocal/nplagent_test.go
+++ b/pkg/agent/nodeportlocal/nplagent_test.go
@@ -36,6 +36,7 @@ import (
 	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/k8s"
 	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
 	npltest "github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/rules/testing"
+	"github.com/vmware-tanzu/antrea/pkg/signals"
 )
 
 func newPortTable(c *gomock.Controller) *portcache.PortTable {
@@ -64,7 +65,6 @@ const (
 
 var kubeClient *k8sfake.Clientset
 var portTable *portcache.PortTable
-var stopCh chan struct{}
 
 func getTestPod() corev1.Pod {
 	return corev1.Pod{
@@ -113,8 +113,9 @@ func getTestSvc() corev1.Service {
 	}
 }
 
-// SetupTest initializes and runs the NPLController.
-func SetupTest(t *testing.T) {
+func TestMain(t *testing.T) {
+	os.Setenv("NODE_NAME", defaultNodeName)
+	kubeClient = k8sfake.NewSimpleClientset()
 	mockCtrl := gomock.NewController(t)
 	portTable = newPortTable(mockCtrl)
 
@@ -122,7 +123,7 @@ func SetupTest(t *testing.T) {
 	informerFactory := informers.NewSharedInformerFactory(kubeClient, resyncPeriod)
 
 	c, _ := InitController(kubeClient, informerFactory, portTable, defaultNodeName)
-	stopCh = make(chan struct{})
+	stopCh := signals.RegisterSignalHandlers()
 
 	go c.Run(stopCh)
 	informerFactory.Start(stopCh)
@@ -131,17 +132,6 @@ func SetupTest(t *testing.T) {
 	// in-between list and watch call of an informer. This is because fake clientset doesn't support watching with
 	// resourceVersion. A watcher of fake clientset only gets events that happen after the watcher is created.
 	informerFactory.WaitForCacheSync(stopCh)
-}
-
-// Teardown closes the channel to stop NPLController.
-func TeardownTest() {
-	close(stopCh)
-}
-
-func TestMain(m *testing.M) {
-	os.Setenv("NODE_NAME", defaultNodeName)
-	kubeClient = k8sfake.NewSimpleClientset()
-	os.Exit(m.Run())
 }
 
 func pollForPodAnnotation(r *require.Assertions, podName string, found bool) (string, error) {
@@ -165,8 +155,6 @@ func pollForPodAnnotation(r *require.Assertions, podName string, found bool) (st
 // Add a Service (proper annotation) and Pod pair, update Service with bad
 // annotation value and check for Pod annotation and port cache entry removal.
 func TestSvcUpdateAnnotation(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)
@@ -214,8 +202,6 @@ func TestSvcUpdateAnnotation(t *testing.T) {
 // Update existing Service with proper annotation to get Pod annotation and port
 // cache entry back, remove annotation from Service.
 func TestSvcRemoveAnnotation(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -244,8 +230,6 @@ func TestSvcRemoveAnnotation(t *testing.T) {
 }
 
 func TestSvcUpdateSelector(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -274,8 +258,6 @@ func TestSvcUpdateSelector(t *testing.T) {
 }
 
 func TestPodUpdateSelectorLabel(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -308,8 +290,6 @@ func TestPodUpdateSelectorLabel(t *testing.T) {
 // Update existing Service with proper annotation to get Pod annotation and port
 // cache entry back, delete Service.
 func TestDeleteSvc(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -346,8 +326,6 @@ func TestDeleteSvc(t *testing.T) {
 // Add a new Pod with fake k8s client and verify that NPL annotation gets updated
 // and an entry gets added in the local port cache.
 func TestPodAdd(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)
@@ -376,8 +354,6 @@ func TestPodAdd(t *testing.T) {
 
 // Test that any update in the Pod container port is reflected in Pod annotation and local port cache.
 func TestPodUpdate(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	var ann map[string]string
 	var nplData []k8s.NPLAnnotation
 	var data string
@@ -414,8 +390,6 @@ func TestPodUpdate(t *testing.T) {
 
 // Make sure that when a Pod gets deleted, corresponding entry gets deleted from local port cache also
 func TestPodDel(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	r := require.New(t)
 
 	err := kubeClient.CoreV1().Pods(defaultNS).Delete(context.TODO(), defaultPodName, metav1.DeleteOptions{})
@@ -430,8 +404,6 @@ func TestPodDel(t *testing.T) {
 
 // Create a Pod with multiple ports and verify that Pod annotation and local port cache are updated correctly
 func TestPodAddMultiPort(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)
@@ -469,8 +441,6 @@ func TestPodAddMultiPort(t *testing.T) {
 // Create Multiple Pods and test that annotations for both the Pods are updated correctly
 // and local port cache is updated accordingly
 func TestAddMultiplePods(t *testing.T) {
-	SetupTest(t)
-	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)

--- a/pkg/agent/nodeportlocal/nplagent_test.go
+++ b/pkg/agent/nodeportlocal/nplagent_test.go
@@ -36,7 +36,6 @@ import (
 	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/k8s"
 	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
 	npltest "github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/rules/testing"
-	"github.com/vmware-tanzu/antrea/pkg/signals"
 )
 
 func newPortTable(c *gomock.Controller) *portcache.PortTable {
@@ -65,6 +64,7 @@ const (
 
 var kubeClient *k8sfake.Clientset
 var portTable *portcache.PortTable
+var stopCh chan struct{}
 
 func getTestPod() corev1.Pod {
 	return corev1.Pod{
@@ -113,9 +113,8 @@ func getTestSvc() corev1.Service {
 	}
 }
 
-func TestMain(t *testing.T) {
-	os.Setenv("NODE_NAME", defaultNodeName)
-	kubeClient = k8sfake.NewSimpleClientset()
+// SetupTest initializes and runs the NPLController.
+func SetupTest(t *testing.T) {
 	mockCtrl := gomock.NewController(t)
 	portTable = newPortTable(mockCtrl)
 
@@ -123,7 +122,7 @@ func TestMain(t *testing.T) {
 	informerFactory := informers.NewSharedInformerFactory(kubeClient, resyncPeriod)
 
 	c, _ := InitController(kubeClient, informerFactory, portTable, defaultNodeName)
-	stopCh := signals.RegisterSignalHandlers()
+	stopCh = make(chan struct{})
 
 	go c.Run(stopCh)
 	informerFactory.Start(stopCh)
@@ -132,6 +131,17 @@ func TestMain(t *testing.T) {
 	// in-between list and watch call of an informer. This is because fake clientset doesn't support watching with
 	// resourceVersion. A watcher of fake clientset only gets events that happen after the watcher is created.
 	informerFactory.WaitForCacheSync(stopCh)
+}
+
+// Teardown closes the channel to stop NPLController.
+func TeardownTest() {
+	close(stopCh)
+}
+
+func TestMain(m *testing.M) {
+	os.Setenv("NODE_NAME", defaultNodeName)
+	kubeClient = k8sfake.NewSimpleClientset()
+	os.Exit(m.Run())
 }
 
 func pollForPodAnnotation(r *require.Assertions, podName string, found bool) (string, error) {
@@ -155,6 +165,8 @@ func pollForPodAnnotation(r *require.Assertions, podName string, found bool) (st
 // Add a Service (proper annotation) and Pod pair, update Service with bad
 // annotation value and check for Pod annotation and port cache entry removal.
 func TestSvcUpdateAnnotation(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)
@@ -202,6 +214,8 @@ func TestSvcUpdateAnnotation(t *testing.T) {
 // Update existing Service with proper annotation to get Pod annotation and port
 // cache entry back, remove annotation from Service.
 func TestSvcRemoveAnnotation(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -230,6 +244,8 @@ func TestSvcRemoveAnnotation(t *testing.T) {
 }
 
 func TestSvcUpdateSelector(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -258,6 +274,8 @@ func TestSvcUpdateSelector(t *testing.T) {
 }
 
 func TestPodUpdateSelectorLabel(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -290,6 +308,8 @@ func TestPodUpdateSelectorLabel(t *testing.T) {
 // Update existing Service with proper annotation to get Pod annotation and port
 // cache entry back, delete Service.
 func TestDeleteSvc(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	a := assert.New(t)
 	r := require.New(t)
 
@@ -326,6 +346,8 @@ func TestDeleteSvc(t *testing.T) {
 // Add a new Pod with fake k8s client and verify that NPL annotation gets updated
 // and an entry gets added in the local port cache.
 func TestPodAdd(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)
@@ -354,6 +376,8 @@ func TestPodAdd(t *testing.T) {
 
 // Test that any update in the Pod container port is reflected in Pod annotation and local port cache.
 func TestPodUpdate(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	var ann map[string]string
 	var nplData []k8s.NPLAnnotation
 	var data string
@@ -390,6 +414,8 @@ func TestPodUpdate(t *testing.T) {
 
 // Make sure that when a Pod gets deleted, corresponding entry gets deleted from local port cache also
 func TestPodDel(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	r := require.New(t)
 
 	err := kubeClient.CoreV1().Pods(defaultNS).Delete(context.TODO(), defaultPodName, metav1.DeleteOptions{})
@@ -404,6 +430,8 @@ func TestPodDel(t *testing.T) {
 
 // Create a Pod with multiple ports and verify that Pod annotation and local port cache are updated correctly
 func TestPodAddMultiPort(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)
@@ -441,6 +469,8 @@ func TestPodAddMultiPort(t *testing.T) {
 // Create Multiple Pods and test that annotations for both the Pods are updated correctly
 // and local port cache is updated accordingly
 func TestAddMultiplePods(t *testing.T) {
+	SetupTest(t)
+	defer TeardownTest()
 	var data string
 
 	a := assert.New(t)

